[Internal] Design Docs: Adds Design Document for Client Telemetry (#3590)
Azure/azure-cosmos-dotnet-v3: docs/observability.md

    OtherLogic --> GetResponse(Get Response for the request)
    SendResponse --> OperationCall

```

## Client Telemetry (Private Preview)

> ### Inroduction

Reviewer: ```suggestion
### Introduction
```

> It is a feature which can be opted-in by setting environment variables. By opting-in this feature, SDK will start sending aggregated telemetry data every 10 minutes to Microsoft.

Author: Not yet and it will take time as compute gateway release is delayed by 1 month. So meanwhile this is the process, later I will change this accordingly.

> It is a feature which can be opted-in by setting environment variables. By opting-in this feature, SDK will start sending aggregated telemetry data every 10 minutes to Microsoft.

Author: I didn't mention intentionally as it won't be of any use for them.

> It is a feature which can be opted-in by setting environment variables. By opting-in this feature, SDK will start sending aggregated telemetry data every 10 minutes to Microsoft.

Reviewer: Why are we documenting this now if customers can't use it? I think we should either include the environment variable and how to set it up or wait until it's ready.

We collect following information as part of this:

> 1. Cache Latencies : Right now, it covers only ```Collection Cache```

Reviewer 2: No need for triple quotes

2. Client System Usage (during an operation) :
    * CPU usage
    * Memory Usage
    * Thread Starvation
    * Network Connections Opened (only TCP Connections)
3. Operation Latencies and Request Units (RUs).
4. Network Request Latencies. (sampled to, top 10 slowest to a replica)

> > Note: We don't collect any PII data as part of this feature. Right now, it is not self opt in, you need to contact us for more information.

Reviewer: If we want to open the private preview for sign ups let's create a sign up form or email they can reach out to. Let's discuss if this is ready to open up to the public

Author: Removed all opt-in story.

### Components

> **Telemetry Job:** This is a background task which collects the data and send it to a Microsoft service every 10 minutes.

Reviewer: "a Microsoft service" is vague. We should explain what the service is and how it's used

```suggestion
**Telemetry Job:** Background task which collects the data and sends it to a Microsoft service every 10 minutes.
```

Author: I intentionally kept it vague. for customer, it doesn't matter what this service is and what this service does (we can address questions if anybody has), it is ajust microsoft service which takes the data.

> **Telemetry Job:** This is a background task which collects the data and send it to a microsoft service every 10 minutes.

Reviewer 2: ```suggestion
**Telemetry Job:** This is a background task which collects the data and send it to a Microsoft service every 10 minutes.
```

> **Collectors:** It is basically, in-memory storage which keeps the telemetry data, collected during an operation. Right now, there are 3 kind of collectors we have i.e 

Reviewer: ```suggestion
**Collectors:** In-memory storage which keeps the telemetry data collected during an operation. There are 3 types of collectors including: 
```

* _Operational Data Collector_: It keeps operation level latencies and request units.

> * _Network Data Collectors_: It keeps all the metrics related to network or TCP calls. It has its own Sampler which sample-in only slowest TCP calls for a particular replica.

Reviewer 2: Why is this one plural but the others singular?

Author: typo. fixed.

            subgraph NetworkDataCollector[Network Data Collector]
                TcpDatapoint(Network Request Datapoint) --> NetworkHistogram[(Histogram)]
                DataSampler(Sampler)
            end
            subgraph DataCollector[Operational Data Collector]
                OpsDatapoint(Operation Datapoint) --> OperationHistogram[(Histogram)]
            end
            subgraph CacheCollector[Cache Data Collector]
                CacheDatapoint(Cache Request Datapoint) --> CacheHistogram[(Histogram)]
            end
        end
        subgraph TelemetryTask[Telemetry Task Every 10 min]
            CacheAccountInfo(Cached Account Properties) --> VMInfo
            VMInfo(Get VM Information) --> CollectSystemUsage
            CollectSystemUsage(Record System Usage Information) --> GetDataFromCollector
        end
        subgraph Processor
            GetDataFromCollector(Fetch Data from Collectors) --> Serializer
            Serializer(Serialize and divide the Payload) --> SendCTOverHTTP(Send Data over HTTP to Service)
        end
        Storage --> |Get Aggregated data|GetDataFromCollector
    end
```

### Benefits

> Enabling this feature provides numerous benefits. The telemetry data collected will allow us to proactively identify and address potential issues before they impact your consumers. This results in a superior support experience and ensures that most issues are resolved before they even arise. In short, customers with this feature enabled can expect a smoother and more reliable experience.

Reviewer 2: Might want to run this by @jcocchi

Reviewer: Do we have plans to proactively address issues? I thought it would be mostly reactive initially. Let's discuss the customer benefits and experience.

Author: this doc is more about the feature only. Right now, it didn't mention how customer can opt -in for this feature.

Reviewer: Can you remove the word "proactively"? I don't think we have plans to do this in the short term. It's okay if we add more benefits in later.

```suggestion
Enabling this feature provides numerous benefits. The telemetry data collected will allow us to identify and address potential issues. This results in a superior support experience and ensures that some issues can even be resolved before they impact your application. In short, customers with this feature enabled can expect a smoother and more reliable experience.
```

### Impact of this feature enabled
* _Latency_: Customer should not see any impact on latency.

> * _Total RPS_: It depends on the infrastructure on which application using SDK is hosted and other factors but on a fully packed Azure Machine and making just point operations, impact should not be more than 10%.

Reviewer: What does "fully packed" mean here? Do we have recommendations or benchmarks? How does operation type change this estimate?

```suggestion
* _Total RPS_: It depends on the infrastructure the application using the SDK is hosted on among other factors. On a fully packed Azure Machine where the SDK is using point operations, impact should not be more than 10%.
```

    OtherLogic --> GetResponse(Get Response for the request)
    SendResponse --> OperationCall

```

## Send telemetry from SDK to service (Private Preview)

### Introduction

> SDK sends aggregated telemetry data every 10 minutes to Microsoft. We collect following information as part of this feature:

Reviewer 3: Replace: Microsoft -> Azure CosmosDB service

> SDK sends aggregated telemetry data every 10 minutes to Microsoft. We collect following information as part of this feature:

Reviewer 3: ```suggestion
When opted-in CosmosDB SDK collects below aggregated telemetry data every 10 minutes to Azure CosmosDB service. 
```

> 1. Cache Latencies : Right now, it covers only Collection Cache

Reviewer 3: ```suggestion
1. Metadata caches (ex: CollectionCache) miss statistics
```

2. Client System Usage (during an operation) :
    * CPU usage
    * Memory Usage
    * Thread Starvation
    * Network Connections Opened (only TCP Connections)
3. Operation Latencies and Request Units (RUs).

> 4. Network Request Latencies. (sampled to, top 10 slowest to a replica)

Reviewer 3: ```suggestion
4. TOP 10 slower network interactons
```

> Note: We don't collect any PII data as part of this feature.

### Limitations
1. AAD Support is not available.

### Components

**Telemetry Job:** Background task which collects the data and sends it to a Microsoft service every 10 minutes.

**Collectors:** In-memory storage which keeps the telemetry data collected during an operation. There are 3 types of collectors including:
* _Operational Data Collector_: It keeps operation level latencies and request units.
* _Network Data Collector_: It keeps all the metrics related to network or TCP calls. It has its own Sampler which sample-in only slowest TCP calls for a particular replica.
* _Cache Data Collector_: It keeps all the cache call latencies. Right now, only collection cache is covered.

> **Get VM Information**: It makes [Azure Instance Metadata](https://learn.microsoft.com/azure/virtual-machines/instance-metadata-service?tabs=windows) call. If customer is not on Azure VM, we won't have this information and customer will see a warning with exception in the Trace Logs (if enabled).

Reviewer 3: ```suggestion
**Get VM Information**: 

- Azure VM: [Azure Instance Metadata](https://learn.microsoft.com/azure/virtual-machines/instance-metadata-service?tabs=windows) call. 
- Non-Azure VM, <Call out that auto generated GUID will be used>

```

> **Processor**: Its resposibility is to get all the data and divide it into small chunks (<2MB) and send each chunk to the Microsoft service.

Reviewer 3: ```suggestion
**Processor**: Its resposibility is to get all the data and divide it into small chunks (<2MB) and send each chunk to the Azure CosmosDB service.
```

        subgraph Storage[In Memory Storage or Collectors]
            subgraph NetworkDataCollector[Network Data Collector]
                TcpDatapoint(Network Request Datapoint) --> NetworkHistogram[(Histogram)]
                DataSampler(Sampler)
            end
            subgraph DataCollector[Operational Data Collector]
                OpsDatapoint(Operation Datapoint) --> OperationHistogram[(Histogram)]
            end
            subgraph CacheCollector[Cache Data Collector]
                CacheDatapoint(Cache Request Datapoint) --> CacheHistogram[(Histogram)]
            end
        end
        subgraph TelemetryTask[Telemetry Task Every 10 min]
            CacheAccountInfo(Cached Account Properties) --> VMInfo
            VMInfo(Get VM Information) --> CollectSystemUsage
            CollectSystemUsage(Record System Usage Information) --> GetDataFromCollector
        end
        subgraph Processor
            GetDataFromCollector(Fetch Data from Collectors) --> Serializer
            Serializer(Serialize and divide the Payload) --> SendCTOverHTTP(Send Data over HTTP to Service)
        end
        Storage --> |Get Aggregated data|GetDataFromCollector
    end
```

> ### Benefits

Reviewer 3: Let's please push the benefits and impact up for visibility.

> Enabling this feature provides numerous benefits. The telemetry data collected will allow us to identify and address potential issues. This results in a superior support experience and ensures that some issues can even be resolved before they impact your application. In short, customers with this feature enabled can expect a smoother and more reliable experience.

Reviewer 3: @jcocchi can you please take a look?

Author: synced offline

### Impact of this feature enabled
* _Latency_: Customer should not see any impact on latency.

> * _Total RPS_: It depends on the infrastructure the application using SDK is hosted on among other factors. On a fully packed Azure Machine (100% CPU usage) where the SDK is using point operations, impact should not be more than 10%.

Reviewer: Can you define "fully packed Azure Machine"? What does 100% CPU usage mean? Don't we not recommend pushing the machine to 100% CPU utilization? I thought high CPU was one of the goals client telemetry is trying to solve

    OtherLogic --> GetResponse(Get Response for the request)
    SendResponse --> OperationCall

```

## Send telemetry from SDK to service (Private Preview)

### Introduction
When opted-in CosmosDB SDK collects below aggregated telemetry data every 10 minutes to Azure CosmosDB service. 

> 1. Operation(CRUD APIs) Latencies and Request Units (RUs).

Reviewer 3: Is it CRUD only? What about query?

Author: Query is also read operation

2. Metadata caches (ex: CollectionCache) miss statistics

> 3. Client System Usage (during an operation) :

Reviewer 3: ```suggestion
3. Host resource usage:
```

    * CPU usage
    * Memory Usage
    * Thread Starvation
    * Network Connections Opened (only TCP Connections)

> 4. TOP 10 slower network interactions

Reviewer 3: ```suggestion
4. TOP 10 slower network interactions per endpoint
```

> Note: We don't collect any PII data as part of this feature.

### Benefits

> Enabling this feature provides numerous benefits. The telemetry data collected will allow us to identify and address potential issues. This results in a superior support experience and ensures that some issues can even be resolved before they impact your application. In short, customers with this feature enabled can expect a smoother and more reliable experience.

Reviewer 3: and more reliable **support** experience
